core/src/main/java/io/questdb/griffin/engine/functions/window/FirstNotNullValueDoubleWindowFunctionFactory.java

```diff
@@ -0,0 +1,766 @@
+                    }
+
+                    int timestampIndex = windowContext.getTimestampIndex();
+
+                    // first_not_null_value() over range between timestamp - rowsLo and timestamp + rowsHi (inclusive)
+                    return new FirstNotNullValueOverRangeFrameFunction(
+                            rowsLo,
+                            rowsHi,
+                            args.get(0),
+                            configuration,
+                            timestampIndex
+                    );
+                }
+            } else if (framingMode == WindowColumn.FRAMING_ROWS) {
+                // between unbounded preceding and [current row | unbounded following]
+                if (rowsLo == Long.MIN_VALUE && (rowsHi == 0 || rowsHi == Long.MAX_VALUE)) {
+                    return new FirstValueDoubleWindowFunctionFactory.FirstValueOverWholeResultSetFunction(args.get(0), true) {
+                        @Override
+                        public String getName() {
+                            return NAME;
+                        }
+                    };
+                } // between current row and current row
+                else if (rowsLo == 0 && rowsLo == rowsHi) {
+                    return new FirstValueDoubleWindowFunctionFactory.FirstValueOverCurrentRowFunction(args.get(0));
+                } // between [unbounded | x] preceding and [y preceding | current row]
+                else {
+                    MemoryARW mem = Vm.getARWInstance(
+                            configuration.getSqlWindowStorePageSize(),
+                            configuration.getSqlWindowStoreMaxPages(),
+                            MemoryTag.NATIVE_CIRCULAR_BUFFER
+                    );
+
+                    return new FirstNotNullValueOverRowsFrameFunction(
+                            args.get(0),
+                            rowsLo,
+                            rowsHi,
+                            mem
+                    );
+                }
+            }
+        }
+
+        throw SqlException.$(position, "function not implemented for given window parameters");
+    }
+
+    // handles first_not_null_value() over (partition by x)
+    // order by is absent so default frame mode includes all rows in the partition
+    static class FirstNotNullValueOverPartitionFunction extends BasePartitionedDoubleWindowFunction {
+
+        public FirstNotNullValueOverPartitionFunction(Map map, VirtualRecord partitionByRecord, RecordSink partitionBySink, Function arg) {
+            super(map, partitionByRecord, partitionBySink, arg);
+        }
+
+        @Override
+        public String getName() {
+            return NAME;
+        }
+
+        @Override
+        public int getPassCount() {
+            return WindowFunction.TWO_PASS;
+        }
+
+        @Override
+        public void pass1(Record record, long recordOffset, WindowSPI spi) {
+            partitionByRecord.of(record);
```

Comment:
the logic is unclear to me, why do we need 2 passes over the window to find first non-null value?